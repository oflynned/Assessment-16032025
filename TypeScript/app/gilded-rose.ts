import { DeprecatedStrategy } from '@/strategies/legacy.strategy';

export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;

  private readonly mode: 'old' | 'new' = 'new';

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  private updateItemQuality(item: Item) {}

  private updateItemSellIn(item: Item) {}

  private handleExpiredItems(item: Item) {}

  updateQuality() {
    if (this.mode === 'old') {
      const strategy = new DeprecatedStrategy();

      for (let i = 0; i < this.items.length; i++) {
        // legacy code lifted from the original implementation with minimal changes
        // to make it work with individual items
        strategy.updateQuality(this.items[i]);
      }

      return this.items;
    }

    for (const item of this.items) {
      // this.updateItemQuality(item);
      // this.updateItemSellIn(item);
      // this.handleExpiredItems(item);
    }

    return this.items;
  }
}
