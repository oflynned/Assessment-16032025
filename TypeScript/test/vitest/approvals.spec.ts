import { GildedRose, Item } from '@/gilded-rose';
import { describe, expect, it } from 'vitest';

describe('Gilded Rose Approval', () => {
  it('should simulate 30 days of updates', () => {
    const items = [
      new Item('+5 Dexterity Vest', 10, 20),
      new Item('Aged Brie', 2, 0),
      new Item('Elixir of the Mongoose', 5, 7),
      new Item('Sulfuras, Hand of Ragnaros', 0, 80),
      new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20),
      new Item('Conjured Mana Cake', 3, 6),
    ];

    const gildedRose = new GildedRose(items);

    const results: string[] = [];

    for (let day = 0; day < 30; day++) {
      results.push(`-------- Day ${day} --------`);
      results.push(JSON.stringify(items, null, 2));
      gildedRose.updateQuality();
    }

    expect(results.join('\n')).toMatchSnapshot();
  });
});
