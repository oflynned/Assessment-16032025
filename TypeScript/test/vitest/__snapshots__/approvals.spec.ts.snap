// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gilded Rose Approval > should simulate 30 days of updates 1`] = `
"-------- Day 0 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 10,
    "quality": 20
  },
  {
    "name": "Aged Brie",
    "sellIn": 2,
    "quality": 0
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": 5,
    "quality": 7
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 15,
    "quality": 20
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": 3,
    "quality": 6
  }
]
-------- Day 1 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 9,
    "quality": 19
  },
  {
    "name": "Aged Brie",
    "sellIn": 1,
    "quality": 1
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": 4,
    "quality": 6
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 14,
    "quality": 21
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": 2,
    "quality": 4
  }
]
-------- Day 2 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 8,
    "quality": 18
  },
  {
    "name": "Aged Brie",
    "sellIn": 0,
    "quality": 2
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": 3,
    "quality": 5
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 13,
    "quality": 22
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": 1,
    "quality": 2
  }
]
-------- Day 3 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 7,
    "quality": 17
  },
  {
    "name": "Aged Brie",
    "sellIn": -1,
    "quality": 3
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": 2,
    "quality": 4
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 12,
    "quality": 23
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": 0,
    "quality": 0
  }
]
-------- Day 4 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 6,
    "quality": 16
  },
  {
    "name": "Aged Brie",
    "sellIn": -2,
    "quality": 4
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": 1,
    "quality": 3
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 11,
    "quality": 24
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -1,
    "quality": 0
  }
]
-------- Day 5 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 5,
    "quality": 15
  },
  {
    "name": "Aged Brie",
    "sellIn": -3,
    "quality": 5
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": 0,
    "quality": 2
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 10,
    "quality": 25
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -2,
    "quality": 0
  }
]
-------- Day 6 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 4,
    "quality": 14
  },
  {
    "name": "Aged Brie",
    "sellIn": -4,
    "quality": 6
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -1,
    "quality": 1
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 9,
    "quality": 27
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -3,
    "quality": 0
  }
]
-------- Day 7 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 3,
    "quality": 13
  },
  {
    "name": "Aged Brie",
    "sellIn": -5,
    "quality": 7
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -2,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 8,
    "quality": 29
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -4,
    "quality": 0
  }
]
-------- Day 8 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 2,
    "quality": 12
  },
  {
    "name": "Aged Brie",
    "sellIn": -6,
    "quality": 8
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -3,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 7,
    "quality": 31
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -5,
    "quality": 0
  }
]
-------- Day 9 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 1,
    "quality": 11
  },
  {
    "name": "Aged Brie",
    "sellIn": -7,
    "quality": 9
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -4,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 6,
    "quality": 33
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -6,
    "quality": 0
  }
]
-------- Day 10 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": 0,
    "quality": 10
  },
  {
    "name": "Aged Brie",
    "sellIn": -8,
    "quality": 10
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -5,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 5,
    "quality": 35
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -7,
    "quality": 0
  }
]
-------- Day 11 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -1,
    "quality": 9
  },
  {
    "name": "Aged Brie",
    "sellIn": -9,
    "quality": 11
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -6,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 4,
    "quality": 38
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -8,
    "quality": 0
  }
]
-------- Day 12 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -2,
    "quality": 7
  },
  {
    "name": "Aged Brie",
    "sellIn": -10,
    "quality": 12
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -7,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 3,
    "quality": 41
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -9,
    "quality": 0
  }
]
-------- Day 13 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -3,
    "quality": 5
  },
  {
    "name": "Aged Brie",
    "sellIn": -11,
    "quality": 13
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -8,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 2,
    "quality": 44
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -10,
    "quality": 0
  }
]
-------- Day 14 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -4,
    "quality": 3
  },
  {
    "name": "Aged Brie",
    "sellIn": -12,
    "quality": 14
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -9,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 1,
    "quality": 47
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -11,
    "quality": 0
  }
]
-------- Day 15 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -5,
    "quality": 1
  },
  {
    "name": "Aged Brie",
    "sellIn": -13,
    "quality": 15
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -10,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": 0,
    "quality": 50
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -12,
    "quality": 0
  }
]
-------- Day 16 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -6,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -14,
    "quality": 16
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -11,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -1,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -13,
    "quality": 0
  }
]
-------- Day 17 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -7,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -15,
    "quality": 17
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -12,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -2,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -14,
    "quality": 0
  }
]
-------- Day 18 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -8,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -16,
    "quality": 18
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -13,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -3,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -15,
    "quality": 0
  }
]
-------- Day 19 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -9,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -17,
    "quality": 19
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -14,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -4,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -16,
    "quality": 0
  }
]
-------- Day 20 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -10,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -18,
    "quality": 20
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -15,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -5,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -17,
    "quality": 0
  }
]
-------- Day 21 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -11,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -19,
    "quality": 21
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -16,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -6,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -18,
    "quality": 0
  }
]
-------- Day 22 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -12,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -20,
    "quality": 22
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -17,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -7,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -19,
    "quality": 0
  }
]
-------- Day 23 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -13,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -21,
    "quality": 23
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -18,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -8,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -20,
    "quality": 0
  }
]
-------- Day 24 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -14,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -22,
    "quality": 24
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -19,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -9,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -21,
    "quality": 0
  }
]
-------- Day 25 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -15,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -23,
    "quality": 25
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -20,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -10,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -22,
    "quality": 0
  }
]
-------- Day 26 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -16,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -24,
    "quality": 26
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -21,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -11,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -23,
    "quality": 0
  }
]
-------- Day 27 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -17,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -25,
    "quality": 27
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -22,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -12,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -24,
    "quality": 0
  }
]
-------- Day 28 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -18,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -26,
    "quality": 28
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -23,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -13,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -25,
    "quality": 0
  }
]
-------- Day 29 --------
[
  {
    "name": "+5 Dexterity Vest",
    "sellIn": -19,
    "quality": 0
  },
  {
    "name": "Aged Brie",
    "sellIn": -27,
    "quality": 29
  },
  {
    "name": "Elixir of the Mongoose",
    "sellIn": -24,
    "quality": 0
  },
  {
    "name": "Sulfuras, Hand of Ragnaros",
    "sellIn": 0,
    "quality": 80
  },
  {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "sellIn": -14,
    "quality": 0
  },
  {
    "name": "Conjured Mana Cake",
    "sellIn": -26,
    "quality": 0
  }
]"
`;
